SR-14635: Casts to NSCopying should not always succeed

tbkka opened the pull request:
The runtime dynamic casting logic explores a variety of strategies for
each cast request.  One of the last options is to wrap the source
in a `__SwiftValue` box so it can bridge to Obj-C.  The previous
code was overly aggressive about such boxing; it performed the boxing
for any source type and only checked to verify that the `__SwiftValue`
box itself was compatible with the destination.

Among other oddities, this results in the behavior discussed
in SR-14635, where any Swift or Obj-C type will always successfully cast
to NSCopying or to NSObjectProtocol because `__SwiftValue` is compatible with NSCopying and NSObjectProtocol.

This is actually two subtly different issues:

* Class types should not be subject to `__SwiftValue` boxing at all.
  Casting class types to class existentials is already handled elsewhere,
  so this function should just reject any source with class type.

* Non-class types should be boxed only when being assigned to
  an AnyObject (an "unconstrained class existential").  If
  the class existential has constraints, it is by definition
  a class-constrained existential which should not receive
  any non-class object.

To solve these, this PR disables `__SwiftValue` boxing in two cases:
1. If the source is a class (reference) type.
2. If the destination has constraints

Resolves SR-14635
Resolves rdar://78224322

tbkka committed:
SR-14635: Casts to NSCopying should not always succeed

The runtime dynamic casting logic explores a variety of strategies for
each cast request.  One of the last options is to wrap the source
in a `__SwiftValue` box so it can bridge to Obj-C.  The previous
code was overly aggressive about such boxing; it performed the boxing
for any source type and only checked to verify that the `__SwiftValue`
box itself was compatible with the destination.

Among other oddities, this results in the behavior discussed
in SR-14635, where any Swift or Obj-C type will always successfully cast
to NSCopying because `__SwiftValue` is compatible with NSCopying.

This is actually two subtly different issues:

* Class types should not be subject to `__SwiftValue` boxing at all.
  Casting class types to class existentials is already handled elsewhere,
  so this function should just reject any source with class type.

* Non-class types should be boxed only when being assigned to
  an AnyObject (an "unconstrained class existential").  If
  the class existential has constraints, it is by definition
  a class-constrained existential which should not receive
  any non-class object.

To solve these, this PR disables `__SwiftValue` boxing in two cases:
1. If the source is a class (reference) type.
2. If the destination has constraints

Resolves SR-14635
Resolves rdar://78224322

mikeash commented on stdlib/public/runtime/DynamicCast.cpp:
```diff
@@ -1514,41 +1514,49 @@ tryCastToClassExistentialViaSwiftValue(
+  /*
+  // I suspect metatypes should also never get boxed here,
+  // but the current test suite fails with this, so this needs
+  // more research.
```
Is it failing on Darwin or just other platforms? Metatypes aren't objects on other platforms, so they'd have to be boxed. Seems like they could avoid boxing on Darwin, but maybe there's some subtlety there.

tbkka replied:
Failing on Darwin, too, which is why I was so surprised.  It's possible this would be a bug in the test suite or even in the spec, but I've not yet had time to work through those particular test cases to figure out what's going on.  It's also possible that I'm just wrong and it's perfectly okay for metatypes to get boxed.  :shrug:

mikeash replied:
Oh wait, it's only class metatypes that are objects on Darwin. The most accurate thing here would probably be to box everything on non-Darwin and box all non-class types on Darwin. That feels kind of weird though.

tbkka replied:
Ah,  yes.  I remember at least one of the failing test was for a struct metatype.  That would explain it.  Best answer is probably to leave it as-is for now.

tbkka replied:
Ack.  No, we have to avoid boxing class metatypes on Darwin.  Otherwise, we risk having those metatypes be unusable when we pass them to Obj-C code.  I'll fix that.

mikeash replied:
Yeah, ObjC will get unhappy if you give it a SwiftValue wrapping a class. So, weird or not, seems like it has to be that way.

tbkka replied:
I fleshed this out to avoid boxing class metatypes on Darwin.  And I also added a casting test case to verify that both struct metatypes and class metatypes can be round-tripped through Any, AnyObject, etc.

tbkka committed:
Avoid boxing class metatypes on Darwin

But continue boxing
 * Non-class metatypes on all platforms
 * All metatypes on non-Darwin platforms

Obj-C interop requires that we do not box class metatypes;
those must be usable as simple pointers when passed to Obj-C.
But no other metatype is object-compatible, so we have to
continue boxing everything else.

tbkka commented:
@swift-ci Please test

tbkka committed:
Split out ObjC-specific test cases

tbkka commented:
@swift-ci Please test
